### PR TITLE
Wiki main page

### LINKED ISSUE #2364: Implement Main Page for osu!wiki
there was this design somewhere but no idea if it's relevant anymore

### PR DESCRIPTION
**Pull request by LiquidPL**
Closes #2364.

---
Adds a customizable frontpage with a search box to the wiki.

### CODE REVIEW

File: resources/assets/coffee/react/wiki-search/suggestion.coffee [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    pathCleaned = @cleanPath @props.suggestion.path, @props.suggestion.title
+
+    div
+      className: osu.classWithModifiers 'wiki-search-suggestions__suggestion', ['highlighted' if @props.highlighted]
+      position: @props.position
+      onMouseEnter: @onMouseEnter
+      onMouseLeave: @onMouseLeave
+      onClick: @onClick
+      @formatTitle titleSplit, queryString
```
Comment (notbakaneko) [inline resources/assets/coffee/react/wiki-search/suggestion.coffee]: text after the matching keyword gets truncated; e.g. `Some Tournament 2018` becomes `Some Tournament` when searching for `Tournament`, `Taiko Wars` truncated to `Taiko`, etc.
Use Elasticsearch to get the highlighting

File: app/Models/Wiki/MainPage.php [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    public function getContent(bool $force = false)
+    {
+        try {
+            $sections = OsuWiki::fetchContent('wiki/'.$this->pagePath());
+        } catch (GitHubNotFoundException $e) {
+            $sections = null;
+        }
+
+        if (present($sections)) {
+            $sections = $this->parseArray(Yaml::parse($sections));
```
Comment (notbakaneko) [inline app/Models/Wiki/MainPage.php]: This can just go into the `try` block

File: app/Libraries/Search/WikiSearch.php [hunk 3]
```diff
@@ -98,10 +102,19 @@ public function getQuery()
...
                 ]]);
+            }
         }
 
         return (new BoolQuery)
             ->must($langQuery)
-            ->must($matchQuery);
+            ->must($matchQuery)
+            ->mustNot(['match' => [
+                'queryable' => false,
```
Comment (notbakaneko) [inline app/Libraries/Search/WikiSearch.php]: probably better to base on document type instead

**Comment by LiquidPL** [comment]
It's probably better to start with a fresh PR in this case.

**Comment by LiquidPL** [comment]
...or not, I need to think this through. I'll keep this close until I figure this out.

**Pull request closed without merging**